### ISSUE TITLE
[Consensus] Non-blocking timestamp processing

### ISSUE THREAD

**Issue by dmitri-perelman** [libra/libra#1521]
## Current state
Some pieces of our code today are dedicated to postponing the vote in case the timestamp of a proposal is "slightly" ahead of the local wall clock time. In this case a node is actually waiting until its wall clock time is getting to the point of a physical clock and sends the vote only after that.

The idea behind waiting prior to voting is to be able to guarantee that **"if a block with timestamp T is committed there are at least f+1 honest validators with physical clock > T at the time of execution"**
This guarantee makes it possible for VM to reason about time-based contracts: there is a reasonable upper bound on the timestamps of the transactions on chain.

## The problem
* The waiting approach is blocking the system, increasing latencies and reducing throughput.
   * A carefully crafted proposals with timestamps that are slightly ahead of the physical clock are going to slow down everything.
   * Even in the benign case with a slight clock skew we might see cases of some validators waiting before voting for the proposals.
* There is a non-trivial amount of code dedicated to waiting in these cases, that might be gone.

## Proposal
We can change the guarantee from:
**"if a block with timestamp T is committed there are at least f+1 honest validators with physical clock > T at the time of execution"**,
to:
**"if a block with timestamp T is committed there are at least f+1 honest validators with physical clock > T - delta at the time of execution"**,

Vote processing changes: if a proposal's timestamp is within the current wall clock + delta, then allowed to vote, otherwise not allowed to vote.
Proposal generation changes: if a parent's timestamp is within the current wall clock + delta, then propose with the parent's timestamp + 1us, otherwise do not propose.

**Comment by mimoo** [comment]
Thanks for writing this. I see one attack with this approach:

* At time `t0`, I propose a block for round 1 with `timestamp = t0 + delta`
* Every honest validator votes for it, because `delta` is the maximum margin of error validators can accept
* At round 2, the leader propose a block with `timestamp = t1 < t0 + delta`.
* No honest node validator for it because the timestamp is not strictly increasing from the previous block.

What about a completely different approach: every validator includes a timestamp in their vote. The timestamp for a block is the average of all votes corrected against f potential fake timestamps (not sure how to do the last part).

**Comment by dmitri-perelman** [comment]
@mimoo we need the timestamps to be signed and we want them to be identical so that we'll be able to generate the QCs / LedgerInfoWithSignatures / etc. 
Note that a proposal **must** know about its parent (because it carries a QC to the parent), so the monotonically increasing timestamp requirement can never be violated.

**Comment by bmaurer** [comment]
> A carefully crafted proposals with timestamps that are slightly ahead of the physical clock are going to slow down everything.

How is this different than sending proposals towards the end of the round?

> Even in the benign case with a slight clock skew we might see cases of some validators waiting before voting for the proposals.

FWIW there's really no legitimate reason for validators to be more than a few milliseconds out of sync. Anything more than that is likely H/W issues which shouldn't strike many validators at the same time.

Anyways, this proposals basically boils down to having timestamps that lag behind real time. I don't feel super strongly either way (though it seems like a bit of a bummer to lose a more accurate timestamp). One thing I will point out -- as structured it seems like if nodes are honest the timestamps will generally be accurate But in byzantine cases the timestamp could be inaccurate by up to delta.

I think that's a bit risky because people might code assuming accurate timestamps. In a hypothetical contract that allowed you to bet on interest rates, imagine the fed announces at 9am. It'd be logical to close betting at exactly 9am. However really you'd need to close at 9am - delta because if a block commits 9:01 am it's possible that the "true" time is 9:01am - delta

**Comment by mimoo** [comment]
> @mimoo we need the timestamps to be signed and we want them to be identical so that we'll be able to generate the QCs / LedgerInfoWithSignatures / etc.


> Note that a proposal must know about its parent (because it carries a QC to the parent), so the monotonically increasing timestamp requirement can never be violated.

**Comment by dmitri-perelman** [comment]
> I think that's a bit risky because people might code assuming accurate timestamps. In a hypothetical contract that allowed you to bet on interest rates, imagine the fed announces at 9am. It'd be logical to close betting at exactly 9am. However really you'd need to close at 9am - delta because if a block commits 9:01 am it's possible that the "true" time is 9:01am - delta

I might miss something but I think the very same problem would happen with the current approach as well: the proposal that is sent by a byzantine validator at 9am-delta can include a timestamp of 9am. As a result the rest of the validators would first wait for 9am and then execute the proposal. It's a matter of whether you "wait and execute", or you "pretend that you waited and execute right away" :).

**Comment by bmaurer** [comment]
> I might miss something but I think the very same problem would happen with the current approach as well: the proposal that is sent by a byzantine validator at 9am-delta can include a timestamp of 9am. As a result the rest of the validators would first wait for 9am and then execute the proposal. It's a matter of whether you "wait and execute", or you "pretend that you waited and execute right away" :).

Because the smart contract would say:

```
if (now() < 9am) { /* allow bet */ }
```

It's OK if a byzantine validator can delay the commit of a transaction, it's not OK if it can, at 9:01 make a transaction appear to commit at 8:59.

(nb: yes, it is possible for a byzantine validator to have two parallel chains that commit conflicting events at 8:59, then at 9:01 decide the favorable chain to commit)

**Comment by mimoo** [comment]
> it's not OK if it can, at 9:01 make a transaction appear to commit at 8:59

you mean you wish it would not be OK, because it is the case in most cryptocurrencies.

**Comment by bmaurer** [comment]
Sorry, my bad on the ordering here. The issue is you can't make a transaction appear to happen at 9:01 until 9:01. A payment channel is a good example here -- if the dispute resolution period ends at 9:01 you can't make it come early.

One way we could potentially stop the clock from falling too far behind (i.e. from somebody proposing at 9:01 a block at 8:59) is to say that if you receive a proposal containing a time T that if T + (round timeout) > local_time() you ignore the proposal / timeout. In other words -- if T is the timestamp in the proposal and tmo is the round timeout, the validator will only vote on the block if the local time when it received the proposal is in the interval [T - tmo, T + tmo].

We could chose if, on top of that, we want to add the restriction "do not vote on a block until local_time > block timestamp". IMHO with Avery's new non-blocking processor suggestion such a condition wouldn't increase the range of things a byzantine validator can do (because they can always give you the proposal at the end of the round) and would be unlikely to cause local performance issues.

**Comment by mimoo** [comment]
> A payment channel is a good example here

I think payment channels get around that by having large periods of times for you to react.

> One way we could potentially stop the clock from falling too far behind

I think this is very close to @dmitri-perelman proposition if I understand correctly? Which sounds like a fair design decision. You want to allow a margin of error around your local time.

**Comment by dmitri-perelman** [comment]
Yup, the proposal is exactly about non-blocking execution of the proposals that fall within a delta of the wall-clock time and rejection of the others.

By the way, I think some of the confusion happens because we didn't specify the way a VM is reading time for its `now()` function (in @bmaurer example). Imho the easiest way to get deterministic execution is to use the `timestamp` field of a proposal. If this is indeed the case:
* We need to update the code because today consensus is not passing the timestamp to the execution.
* If a block timestamp determines the time for VM execution, then "wait and execute" policy doesn't make sense at all: indeed, what's the point of waiting for `20ms` if the VM input doesn't change?

**Comment by bmaurer** [comment]
> By the way, I think some of the confusion happens because we didn't specify the way a VM is reading time for its now() function (in @bmaurer example). Imho the easiest way to get deterministic execution is to use the timestamp field of a proposal. If this is indeed the case:

Agree that transactions proposed as part of a block B should appear to happen at the time B.timestamp, I think as part of the work @runtian-zhou is doing we could pass this timestamp to a "clock" smart contract that would update the clock at every block.

> If a block timestamp determines the time for VM execution, then "wait and execute" policy doesn't make sense at all: indeed, what's the point of waiting for 20ms if the VM input doesn't change?

We could execute right away, we just wouldn't send the vote until the timestamp had passed. The benefit of this is that a QC for a block cannot be formed until at least f+1 honest nodes believe that B.timestamp has passed. Maybe it's a fair point that this isn't of much practical benefit to the user. Let's say a payment channel expires at time T. The round ends at time TMO, T < TMO. We are the proposer for this round and want to prevent a claim on the channel from coming in before time TMO.

We could make a proposal and set timestamp = TMO+1. In the current setup we'd wait until TMO+1 to vote on the block. However the proposer could just as well have waited until TMO+1 to make the proposal.

So if we take the approach where we discard blocks where B.timestamp > (RoundStart + RoundTimeout) we get the effect of discarding any block that we would have timed out under the original strategy. While it's possible that we will vote for a block that is slightly in the future relative to our local clock, a malicious proposer had a "lock" on that round anyways. I.e. they could have just as easily delayed their proposal and achieved the later timestamp. (n.b.: this is a slightly tighter bound than LocalNow() + RoundTimeout)

**Comment by dmitri-perelman** [comment]
Exactly!

**Comment by dmitri-perelman** [comment]
Ok, discussed it with @aching who is the current owner of timestamp processing.
@aching doesn't particularly like this idea because it might make it harder for the lightweight clients to reason about the transactions that have been committed given the client's local time (e.g., if a client knows its physical local time it `t` it might want to be sure that the timestamp of the latest committed transaction is less than `t` and not less than `t+delta`). 
Reassigning the issue to @aching: feel free to close it!

**Comment by kphfb** [comment]
@aching a client should never rely on its local timestamp for this.  It can only truly know if a transaction has been committed or omitted by querying the chain and seeing a committed timestamp on chain.  The client's local time should never be used for determining if a transaction has been committed - in fact, that is not an accurate way to determine a transaction commitment and would be misleading if we implied that to be the case

**Comment by aching** [comment]
First of all, thanks to @dmitri-perelman for bringing this up.

> The waiting approach is blocking the system, increasing latencies and reducing throughput.
A carefully crafted proposals with timestamps that are slightly ahead of the physical clock are going to slow down everything.
> Even in the benign case with a slight clock skew we might see cases of some validators waiting before voting for the proposals.

This is equivalent to a malicious proposer that sends a proposal after a delay in the round.  In the worst case, it can simply not send a proposal at all, forcing a round timeout.

Slowing everything down if unfortunately an artifact of today's implementation.  We will be discussing an improvement in our message processing that would allow parallel message checking, timeout/vote processing, during the potential wait.

Additionally, it will be fairly easy to figure out if there is a large variance of clocks between validators and suggest that they fix it - the counters are in place for this.  Validators that have clock skew beyond a specific tolerance are considered malicious.

Finally, in testnet and other cluster deployments we have not observed any actual waits in the counters since message latency (~30ms) seems to exceed any clock skew between validators.

> There is a non-trivial amount of code dedicated to waiting in these cases, that might be gone.

So even if we add a fixed delta to the property, unfortunately I don't think the code simplifies.  The checks would need to include the delta but would still exist.  If we add a variable delta (e.g. round timeout), delta is now bound to the round timeout that is supposed to increase faster than any skew that occurs during the round.  A variable delta will be harder to reason about for clients.  We can consider today's system to have a delta of 0.

> Proposal
> We can change the guarantee from:
> "if a block with timestamp T is committed there are at least f+1 honest validators with physical clock > T at the time of execution",
> to:
> "if a block with timestamp T is committed there are at least f+1 honest validators with physical > clock > T - delta at the time of execution",

As mentioned above if delta = 0, they are equivalent.  We expect that clients can tell if they are close to the tail of the blockchain by checking the highest blockchain time against they own clock, adding more delta just introduces more uncertainty around that.  If delta is small it doesn't matter, but if delta was large (e.g. minutes or hours), this would make it harder for clients.

In summary, at this time I would prefer to avoid adding any delta != 0 that changes the property for clients without a tangible gain.

**Comment by aching** [comment]
> @aching a client should never rely on its local timestamp for this. It can only truly know if a transaction has been committed or omitted by querying the chain and seeing a committed timestamp on chain. The client's local time should never be used for determining if a transaction has been committed - in fact, that is not an accurate way to determine a transaction commitment and would be misleading if we implied that to be the case

@kphfb, I think something was lost in translation.  =) I agree with everything you wrote.  About client time syncing, I was saying that a client that syncs to a ledger info that contains a committed blockchain time can reason about how close the client is to the end of the blockchain based on a clock that is synced to the global time (e.g. its local clock) without worrying about a delta.  Obviously the smaller delta is, the bounds are tighter and as delta increases the bounds are looser.  For instance if I believe my local clock is synced to the global time and I see a commit with a clock 100 ms in the past, I am within 100 ms of the blockchain global time.  If I add a delta of 10 seconds for instance, I am within 10.1 seconds of the blockchain global time.

**Comment by aching** [comment]
Discussed with @kphfb and @dmitri-perelman and it makes sense to close this for now.
